File: src/main/java/com/hanocybous/model/ComplexTask.java
package com.hanocybous;

import org.jetbrains.annotations.NotNull;

public final class ComplexTask extends Task {

    public ComplexTask(int id, String name, int mamaId) {
        super (id,name,mamaId);
    }

    public void addSubTask(Task subTask) {
        if(this.getStart() == -1 && this.getEnd() == -1) {
            setStart(subTask.getStart());
            setEnd(subTask.getEnd());
        }
        else  {
            if(getStart() > subTask.getStart() ) {
                setStart(subTask.getStart());
            }
            if(getEnd() < subTask.getEnd()) {
                setEnd(subTask.getEnd());
            }
        }
        addCost(subTask.getCost());
    }

    public void removeSubTask(@NotNull Task subTask) {
        if(this.getStart() == subTask.getStart() && this.getEnd() == subTask.getEnd()) {
            setStart(-1);
            setEnd(-1);
        }
        else {
            if(getStart() == subTask.getStart()) {
                setStart(subTask.getEnd());
            }
            if(getEnd() == subTask.getEnd()) {
                setEnd(subTask.getStart());
            }
        }
        removeCost(subTask.getCost());
    }

    private void removeCost(double cost) {
        addCost(getCost() - cost);
    }
}

File: src/main/java/com/hanocybous/model/ComplexTask.kt
package com.hanocybous.model

import org.jetbrains.annotations.NotNull

class ComplexTask(
    id: Int,
    name: String,
    mamaId: Int
) : Task(id, name, mamaId) {

    private var subTasks: MutableList<Task> = ArrayList()

    override fun addSubTask(subTask: Task) {
        if (this.start == -1 && this.end == -1) {
            start = subTask.start
            end = subTask.end
        } else {
            if (start > subTask.start) {
                start = subTask.start
            }
            if (end < subTask.end) {
                end = subTask.end
            }
        }
        addCost(subTask.cost)
    }

    fun removeSubTask(@NotNull subTask: Task) {
        removeCost(subTask.cost)
        start = -1
        end = -1
        for (task in subTasks) {
            if (task.start != -1 && task.end != -1) {
                if (start == -1 || start > task.start) {
                    start = task.start
                }
                if (end == -1 || end < task.end) {
                    end = task.end
                }
            }
        }
    }

    private fun getSubTasks(): List<Task> {
        return subTasks
    }

    private fun removeCost(cost: Double) {
        addCost(cost - cost)
    }

    fun setSubTasks(subTasks: MutableList<Task>) {
        this.subTasks = subTasks
    }

}